pkg/kubernetes/client.go

```diff
@@ -224,6 +225,62 @@ func (c *RawClient) createOrReplaceObject(object runtime.RawExtension, plan bool
 	return nil
 }
 
+// Delete attempts to delete the Kubernetes resources in the provided manifest,
+// or do nothing if they do not exist.
+func (c *RawClient) Delete(manifest []byte) error {
+	objects, err := NewRawExtensions(manifest)
+	if err != nil {
+		return err
+	}
+	for _, object := range objects {
+		if err := c.deleteObject(object); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *RawClient) deleteObject(object runtime.RawExtension) error {
+	resource, err := c.NewRawResource(object)
+	if err != nil {
+		return err
+	}
+	status, err := resource.Delete()
+	if err != nil {
+		return err
+	}
+	if status != "" {
+		logger.Info(status)
+	}
+	return nil
+}
+
+// Exists checks if the Kubernetes resources in the provided manifest exist or
+// not, and returns a map[<namespace>]map[<name>]bool to indicate each
+// resource's existence.
+func (c *RawClient) Exists(manifest []byte) (map[string]map[string]bool, error) {
```

Comment:
I think this function does too much. Can we do the parsing and existence checks separately?

Also, do we really need to check for existence of each element? Can't we just try to create all the resources in a manifest and do the delete-create on a case-by-case basis? 

By doing that we can piggyback on the manifest parsing done by `CreateOrReplace`, right now we are parsing the manifests twice.

Alternatively we can do the parsing ourselves and make `CreateOrReplace` operate on RawResources.

Reply:
The more I think about it, the better it feels for `Exists()` and `CreateOrReplace()` to operate on resources and not on full manifests:

1. We decouple parsing and operations (which makes the code easier to read)
2. It simplifies handling of individual resources as opposed to managing a chunk of resources living in a manifest.

Reply:
Oh, Actually I saw you already create `Exists()` for resources. That should make it easier

Reply:
How about `CheckAllResourcesExist(manifest []byte) (bool, error)`? That would simplify things and is rather sufficient for current usage.

Reply:
Having the parsing and actual operations done separately would mean changing the API of `RawClient` so that `CreateOrReplace`, `Exists` and `Delete` take `[]runtime.RawExtension` in, instead of `[]byte` like it does now, hence exposing the implementation details a bit more.
Do we really want that? 🤔 

(I personally don't mind parsing twice for the sake of, arguably, keeping this a bit simpler, at the cost of redundant work and associated CPU and I/O waste, given this is still pretty light in absolute. But I also don't really like the abstractions we have now. So I'm not against refactoring this either, but maybe just not now... WDYT? ⌚️)

> Can't we just try to create all the resources in a manifest and do the delete-create on a case-by-case basis?

Sorry, but not sure I understand what you have in mind there? 😅

Reply:
`CheckAllResourcesExist` isn't really what I need here though. I only need to check if only field in one `Secret` in one file is present or not. Are we 100% sure nothing adds any other field to that `Secret`?

Reply:
Given this is all about the fact that Flux manifests contain an empty secret, and going forward we have https://github.com/weaveworks/eksctl/issues/1213 to close on, I'd leave this as an inevitable consequence of having to handle the secret in a special way.

Reply:
Merged. We need to move forward. I'm happy to refine, refactor, clean-up, etc. later once clear how we want to proceed. 👍